skorch/callbacks/training.py

```diff
@@ -761,3 +762,64 @@ def on_train_end(self, net, **kwargs):
         self.checkpoint_.save_model(net)
         self.checkpoint_._sink("Final checkpoint triggered", net.verbose)
         return self
+
+
+class InputShapeSetter(Callback):
+    """Sets the input dimension of the PyTorch module to the input dimension
+    of the training data. This can be of use when using a skorch model within
+    a sklearn pipeline and grid-searching feature transformers (or using
+    feature selection methods).
+
+    Basic usage:
+    >>> class MyModule(torch.nn.Module):
+    ...     def __init__(self, input_dim=1):
+    ...         super().__init__()
+    ...         self.layer = torch.nn.Linear(input_dim, 3)
+    ... # ...
+    >>> X1 = np.zeros(100, 5)
+    >>> X2 = np.zeros(100, 3)
+    >>> y = np.zeros(100)
+    >>> net = NeuralNetClassifier(MyModule, callbacks=[InputShapeSetter()])
+    >>> net.fit(X1, y) # sets self.module_ = module(input_dim=5)
+    >>> net.fit(X2, y) # sets self.module_ = module(input_dim=3)
+
+    Parameters
+    ----------
+    module_name : str (default='module')
+      Only needs change when you are using more than one module in your
+      skorch model (e.g., in case of GANs)
+
+    param_name : str (default='input_dim')
+      The parameter name is the parameter your model uses to define the
+      input dimension in its ``__init__`` method.
+
+    input_dim_fn : callable, None (default=None)
+      In case your ``X`` value is more complex and deriving the input
+      dimension is not as easy as ``X.shape[1]`` you can pass a callable
+      to this parameter which takes ``X`` and returns the input dimension.
+    """
+    def __init__(
+        self,
+        module_name='module',
+        param_name='input_dim',
+        input_dim_fn=None,
+    ):
+        self.module_name = module_name
+        self.param_name = param_name
+        self.input_dim_fn = input_dim_fn
+
+    def get_input_dim(self, X):
+        if self.input_dim_fn is not None:
+            return self.input_dim_fn(X)
+        return X.shape[1]
+
+    def on_train_begin(self, net, X, y):
+        params = net.get_params()
+        input_dim = self.get_input_dim(X)
+        param_name = f'{self.module_name}__{self.param_name}'
+
+        if params.get(param_name, None) == input_dim:
+            return
```

Comment:
```suggestion
            return self
```

Reply:
Are you sure about this? We do not adhere to this throughout skorch and this is not defined in the base signature.